src/plugins/ui_feature/selection_input.ts

```diff
@@ -75,14 +81,36 @@ export class SelectionInputPlugin extends UIPlugin implements StreamCallbacks<Se
   }
 
   handleEvent(event: SelectionEvent) {
-    const inputSheetId = this.activeSheet;
-    const sheetId = this.getters.getActiveSheetId();
-    const zone = event.anchor.zone;
-    const range = this.getters.getRangeFromZone(
-      sheetId,
-      event.options.unbounded ? this.getters.getUnboundedZone(sheetId, zone) : zone
-    );
-    this.add([this.getters.getSelectionRangeString(range, inputSheetId)]);
+    if (this.focusedRangeIndex === null) {
+      return;
+    }
+
+    const inputSheetId = this.inputSheetId;
+    const activeSheetId = this.getters.getActiveSheetId();
+    const zone = event.options.unbounded
+      ? this.getters.getUnboundedZone(activeSheetId, event.anchor.zone)
+      : event.anchor.zone;
+    const range = this.getters.getRangeFromZone(activeSheetId, zone);
+
+    const willAddNewRange =
+      event.mode === "newAnchor" &&
+      !this.inputHasSingleRange &&
+      this.ranges[this.focusedRangeIndex].xc.trim() !== "";
+
+    if (willAddNewRange) {
+      const xc = this.getters.getSelectionRangeString(range, inputSheetId);
+      this.insertNewRange(this.ranges.length, [xc]);
+      this.focusLast();
+    } else {
+      let parts = range.parts;
+      const previousXc = this.ranges[this.focusedRangeIndex].xc.trim();
+      if (previousXc) {
+        parts = RangeImpl.getRangeParts(previousXc, toUnboundedZone(previousXc));
```

Comment:
`getRangeFromXc(...).parts` ? Let's avoid using RangeImpl when not absolutely necessary.